### PR TITLE
Added Timeswap TVL

### PR DESCRIPTION
**Pull request by emmanuelantony2000**
##### Twitter Link:

[https://twitter.com/TimeswapLabs](https://twitter.com/TimeswapLabs)

##### List of audit links if any:

Code4rena has completed our audit. We are waiting for the final report and will share it soon.

##### Website Link:

[https://timeswap.io](https://timeswap.io)

##### Logo (High resolution, preferably in .svg and .png, for application on both white and black backgrounds. Will be shown with rounded borders):

Logo_Wordmark_Dark
Logo_Wordmark_White

##### Current TVL:

Screenshot 2022-04-18 at 9 13 01 AM
Current TVL is $209.11 k
(Ran this script on 3:45AM UTC 18 April 2022)

##### Chain:

- Polygon

##### Coingecko ID (so your TVL can appear on Coingecko): (https://api.coingecko.com/api/v3/coins/list)

Timeswap hasn't yet issued our DAO Token.

##### Coinmarketcap ID (so your TVL can appear on Coinmarketcap): (https://api.coinmarketcap.com/data-api/v3/map/all?listing_status=active,inactive,untracked&start=1&limit=10000)

Timeswap hasn't yet issued our DAO Token.

##### Short Description (to be shown on DefiLlama):

Timeswap is the first fully permissionless, oracle-less, non-liquidatable, fixed maturity lending & borrowing protocol.

##### Token address and ticker if any:

Timeswap hasn't yet issued our DAO Token.

##### Category (full list at https://defillama.com/categories) *Please choose only one:

[Lending](https://defillama.com/protocols/Lending)

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):

Not using any Oracle.

##### forkedFrom (Does your project originate from another project):

None.

##### methodology (what is being counted as tvl, how is tvl being calculated):

Call the balanceOf function for tokens on pair addresses.

### CODE REVIEW

**Comment by llamatester** [comment]
Error while running adapter at projects/timeswap/index.js: 
        
 
  
    
    
    Error: failed to meet quorum (method="getLogs", params={"filter":{"fromBlock":"0x19bdf2b","toBlock":"0x19be36c","address":"0x37938633629B2D260c6dF2e6d62a7E8CfEe097c5","topics":["0x58e59580cc0153b14c553b340b7d7ae6ab3df9747aa39f82941816c5db3869f2"]}}, results=[{"weight":1,"start":1650339157086,"error":{"reason":"timeout","code":"TIMEOUT","requestBody":"{\"method\":\"eth_getLogs\",\"params\":[{\"fromBlock\":\"0x19bdf2b\",\"toBlock\":\"0x19be36c\",\"address\":\"0x37938633629b2d260c6df2e6d62a7e8cfee097c5\",\"topics\":[\"0x58e59580cc0153b14c553b340b7d7ae6ab3df9747aa39f82941816c5db3869f2\"]}],\"id\":542,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","timeout":120000,"url":"https://polygon-rpc.com/"}},{"weight":1,"start":1650339157837,"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"error\":{\"message\":\"Blockheight too far in the past: eth_getLogs. Range of blocks allowed for your plan: 1000\",\"code\":-32005},\"jsonrpc\":\"2.0\",\"id\":376}","error":{"code":-32005},"requestBody":"{\"method\":\"eth_getLogs\",\"params\":[{\"fromBlock\":\"0x19bdf2b\",\"toBlock\":\"0x19be36c\",\"address\":\"0x37938633629b2d260c6df2e6d62a7e8cfee097c5\",\"topics\":[\"0x58e59580cc0153b14c553b340b7d7ae6ab3df9747aa39f82941816c5db3869f2\"]}],\"id\":376,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"https://rpc-mainnet.maticvigil.com/"}}], provider={"_isProvider":true,"_events":[],"_emitted":{"block":-2},"disableCcipRead":false,"formatter":{"formats":{"transaction":{},"transactionRequest":{},"receiptLog":{},"receipt":{},"block":{},"blockWithTransactions":{},"filter":{},"filterLog":{}}},"anyNetwork":false,"_network":{"name":"polygon","chainId":137},"_maxInternalBlockNumber":27308496,"_lastBlockNumber":-2,"_maxFilterBlockRange":10,"_pollingInterval":4000,"_fastQueryDate":1650338414708,"providerConfigs":[{"provider":{"_isProvider":true,"_events":[],"_emitted":{"block":-2},"disableCcipRead":false,"formatter":{"formats":{"transaction":{},"transactionRequest":{},"receiptLog":{},"receipt":{},"block":{},"blockWithTransactions":{},"filter":{},"filterLog":{}}},"anyNetwork":false,"_network":{"name":"polygon","chainId":137},"_maxInternalBlockNumber":27308857,"_lastBlockNumber":-2,"_maxFilterBlockRange":10,"_pollingInterval":4000,"_fastQueryDate":1650339157264,"connection":{"url":"https://polygon-rpc.com/"},"_nextId":543,"_eventLoopCache":{"eth_blockNumber":null},"_internalBlockNumber":{},"_fastBlockNumber":27308857,"_fastBlockNumberPromise":{}},"priority":0,"stallTimeout":750,"weight":1},{"provider":{"_isProvider":true,"_events":[],"_emitted":{"block":27308500},"disableCcipRead":false,"formatter":{"formats":{"transaction":{},"transactionRequest":{},"receiptLog":{},"receipt":{},"block":{},"blockWithTransactions":{},"filter":{},"filterLog":{}}},"anyNetwork":false,"_network":{"name":"polygon","chainId":137},"_maxInternalBlockNumber":27308860,"_lastBlockNumber":27308500,"_maxFilterBlockRange":10,"_pollingInterval":4000,"_fastQueryDate":1650339157966,"connection":{"url":"https://rpc-mainnet.maticvigil.com/"},"_nextId":377,"_eventLoopCache":{"eth_blockNumber":null},"_internalBlockNumber":{},"_poller":null,"_bootstrapPoll":null,"_fastBlockNumber":27308860,"_fastBlockNumberPromise":{}},"priority":1,"stallTimeout":750,"weight":1}],"quorum":1,"_highestBlockNumber":27308496,"_internalBlockNumber":{},"_fastBlockNumber":27308496,"_fastBlockNumberPromise":{}}, code=SERVER_ERROR, version=providers/5.6.2)
        at Logger.makeError (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/logger/lib/index.js:233:21)
        at Logger.throwError (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/logger/lib/index.js:242:20)
        at FallbackProvider.<anonymous> (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:682:54)
        at step (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:48:23)
        at Object.next (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:29:53)
        at fulfilled (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:20:58) {
      reason: 'failed to meet quorum',
      code: 'SERVER_ERROR',
      method: 'getLogs',
      params: {
        filter: {
          fromBlock: '0x19bdf2b',
          toBlock: '0x19be36c',
          address: '0x37938633629B2D260c6dF2e6d62a7E8CfEe097c5',
          topics: [Array]
        }
      },
      results: [
        { weight: 1, start: 1650339157086, error: [Error] },
        { weight: 1, start: 1650339157837, error: [Error] }
      ],
      provider: FallbackProvider {
        _isProvider: true,
        _events: [],
        _emitted: { block: -2 },
        disableCcipRead: false,
        formatter: Formatter { formats: [Object] },
        anyNetwork: false,
        _network: { name: 'polygon', chainId: 137 },
        _maxInternalBlockNumber: 27308496,
        _lastBlockNumber: -2,
        _maxFilterBlockRange: 10,
        _pollingInterval: 4000,
        _fastQueryDate: 1650338414708,
        providerConfigs: [ [Object], [Object] ],
        quorum: 1,
        _highestBlockNumber: 27308496,
        _internalBlockNumber: Promise { [Object] },
        _fastBlockNumber: 27308496,
        _fastBlockNumberPromise: Promise { 27308496 }
      }
    }

**Comment by emmanuelantony2000** [comment]
We faced the same error initially, providing a custom RPC for polygon in the `.env` solved the error for us.

File: projects/timeswap/index.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+function chainTvl(chain) {
+  return async (timestamp, _ethBlock, chainBlocks) => {
+    const block = await getBlock(timestamp, chain, chainBlocks);
+
+    const logs = [];
+
+    for (const factory of factories[chain]) {
+      logs.push(
+        ...(
+          await sdk.api.util.getLogs({
```
Comment (g1nt0ki) [inline projects/timeswap/index.js:28]: @emmanuelantony2000 I think this is the issue, is there a way to get pairs some other way. querying events is painful and tends to break a lot

Comment (emmanuelantony2000) [reply projects/timeswap/index.js:28]: @g1nt0ki apologies for the delayed response, we have set up a subgraph to query the pairs. We can use that.

Comment (g1nt0ki) [reply projects/timeswap/index.js:28]: > @g1nt0ki apologies for the delayed response, we have set up a subgraph to query the pairs. We can use that.

Sounds great, let me know if you need any assistance with changing the adapter to use subgraph

**Comment by waverune** [comment]
Update: 
The latest round of audit has been done by Code4rena.
Here's the Audit Report  :  https://code4rena.com/reports/2022-03-timeswap/

**Comment by g1nt0ki** [comment]
I have updated the info in our db, it should appear in your listing in the next few hours

**Comment by waverune** [comment]
> I have updated the info in our db, it should appear in your listing in the next few hours
@g1nt0ki  the updated info isnt reflected on the [page](https://defillama.com/protocol/timeswap) yet

**Comment by g1nt0ki** [comment]
@waverune it should be fixed now, can you verify?

**Comment by waverune** [comment]
roger that! resolved @g1nt0ki

**Pull request merged**